services/autoscaling/src/simcore_service_autoscaling/dynamic_scaling_core.py

```diff
@@ -1,68 +1,203 @@
-        app,
-        monitored_nodes,
-        cluster_total_resources,
-        cluster_used_resources,
-        pending_tasks,
+    # get the corresponding ec2 instance data
+    # NOTE: some might be in the process of terminating and will not be found
+    ec2_client = get_ec2_client(app)
+    drained_empty_ec2_instances = await asyncio.gather(
+        *(
+            ec2_client.get_running_instance(
+                app_settings.AUTOSCALING_EC2_INSTANCES,
+                tag_keys=[
+                    "io.simcore.autoscaling.version",
+                ],
+                instance_host_name=node.Description.Hostname,
+            )
+            for node in drained_empty_nodes
+            if node.Description and node.Description.Hostname
+        ),
+        return_exceptions=True,
     )
 
+    terminateable_nodes: list[tuple[Node, EC2InstanceData]] = []
+    for node, ec2_instance_data in zip(
+        drained_empty_nodes, drained_empty_ec2_instances
+    ):
+        if isinstance(ec2_instance_data, Ec2InstanceNotFoundError):
+            # skip if already terminating
+            continue
+        # NOTE: AWS price is hourly based (e.g. same price for a machine used 2 minutes or 1 hour, so we wait until 55 minutes)
+        elapsed_time_since_launched = (
+            datetime.utcnow().replace(tzinfo=timezone.utc)
+            - ec2_instance_data.launch_time
+        )
+        elapsed_time_since_full_hour = elapsed_time_since_launched % timedelta(hours=1)
+        if (
+            elapsed_time_since_full_hour
+            >= app_settings.AUTOSCALING_EC2_INSTANCES.EC2_INSTANCES_TIME_BEFORE_TERMINATION
+        ):
+            # let's terminate that one
+            terminateable_nodes.append((node, ec2_instance_data))
+    if terminateable_nodes:
+        logger.info(
+            "the following nodes were found to be terminateable: '%s'",
+            f"{[node.Description.Hostname for node,_ in terminateable_nodes if node.Description]}",
+        )
+    return terminateable_nodes
+
+
+async def _try_scale_down_cluster(app: FastAPI, monitored_nodes: list[Node]) -> None:
+    # 2. once it is in draining mode and we are nearing a modulo of an hour we can start the termination procedure (parametrize this)
+    # NOTE: the nodes that were just changed to drain above will be eventually terminated on the next iteration
+    if terminateable_nodes := await _find_terminateable_nodes(app, monitored_nodes):
+        await asyncio.gather(
+            *(
+                get_ec2_client(app).terminate_instance(ec2_instance_data)
+                for _, ec2_instance_data in terminateable_nodes
+            )
+        )
+        logger.info(
+            "terminated the following machines: '%s'",
+            f"{[node.Description.Hostname for node,_ in terminateable_nodes if node.Description]}",
+        )
+        # since these nodes are being terminated, remove them from the swarm
+        await utils_docker.remove_nodes(
```

Comment:
What if this call fails? Is this somehow handled in a way that the autoscaling continuous in a proper way? Do we need an alert and human intervention in this case?
I have seen docker swarm struggle managing it's nodes, for example when we started an autoscaled machine fro a AMI image where the swarm was still joined, i.e. duplicating swarm members.

Reply:
ok, so the problem you mention here I would describe as very unlikely to happen here. we did join a duplicate of a machine that was already in the swarm with exactly the same node ID... but let's say we do it:
- we have a swarm with nodes
- we take one of the node, and create an AMI out of it
- the autoscaling is started with that AMI
- it will then create nodes that have the same ID as another node already in the swarm
- the swarm will sometimes report the original node, sometimes the created one (as we saw together)
- then the autoscaling will create more nodes still using the same AMI, and they will never be added to the list, probably the swarm will eventually not work anymore at some point

Now what can I do from the autoscaling service point of view? 
1. I do not think I can double check that your AMI is not already connected to a swarm can I? 
2. Maybe I could run ```docker swarm leave``` as a preventive measure as instance bootup script , but it might also corrupt the swarm 
3. The best safeguard I could see, is that you have a test when the AMI is created...

If the call fails, the background task will catch it and restart the task.